feat(VR): enable volumetric lighting

alandtse opened the pull request:
Ready to merge after prereqs.

- [x] https://github.com/doodlum/skyrim-community-shaders/pull/590
- [x] https://github.com/doodlum/skyrim-community-shaders/pull/586
- [x] https://github.com/doodlum/skyrim-community-shaders/pull/585
- [x] https://github.com/doodlum/skyrim-community-shaders/pull/582.

Testing for new bugs.

Known bugs

- [x] bUseHiddenAreaMesh disabled has bad stencil. - Not reproducible on Quest 3. Looks fine for my testing.
- [x] PureDark upscaler ultra performance results in flickering (likely won't fix)
- [x] Disabling dynamic resolution results in bad stencil at edges of eyes. Restart after making this change. (likely won't fix)

alandtse commented:
Current issue is the [VLTex ](https://github.com/doodlum/Skyrim-Community-Shaders/blob/dev/package/Shaders/ISVolumetricLightingGenerateCS.hlsl) only renders something appropriate for the left eye. This results in mismatches in the right eye. Since VL is very subtle, it's hard to notice. Attempts to generate VLTex in stereo are ongoing.

![image](https://github.com/user-attachments/assets/40f153f8-17c6-4c75-a75b-96750fb52f11)
Notice the shaping matches the arch on the left eye, but is not placed properly on the right.
![image](https://github.com/user-attachments/assets/1bf4409a-bc55-4790-b791-e764a626d3a5)

alandtse committed:
feat(VR): enable volumetric lighting

alandtse committed:
fix: fix stereo access to VLTex

VLTex is still mono and causes artifacts, but this is a correct step
once it is stereo.
https://github.com/doodlum/Skyrim-Community-Shaders/pull/565#issuecomment-2381750427

alandtse committed:
fix: fix stereouv access

alandtse commented:
Ok stereo is fixed. Seeing some artifacts with latest dev in dynres though.

alandtse committed:
fix: fix WindInput

FlayaN commented:
Did you see that your previous prereq to this PR was reverted? https://github.com/doodlum/skyrim-community-shaders/commit/398f552d19d0229fc8b05ce183e40d2778142db0

alandtse commented:
No biggie. The improvements were marginal and honestly not something that is too important to me. My rebase should have removed them but I'll verify again.

alandtse committed:
feat: add initial RE for ISFullScreenVR

alandtse committed:
fix: fix previous depth calculation in dynres

The depth stencil from ISDepthBufferCopy is actually full size and not
scaled with dynres enabled.

alandtse committed:
feat: add dynres status to state

alandtse committed:
fix: fix dynres for volumetric lighting

alandtse commented:
Dynres is fixed.

alandtse commented:
Has prereq prs.
https://github.com/doodlum/skyrim-community-shaders/pull/587
https://github.com/doodlum/skyrim-community-shaders/pull/586
https://github.com/doodlum/skyrim-community-shaders/pull/585
https://github.com/doodlum/skyrim-community-shaders/pull/582

alandtse committed:
chore: fix typo

alandtse commented on package/Shaders/ISApplyVolumetricLighting.hlsl:
```diff
@@ -48,14 +57,25 @@ PS_OUTPUT main(PS_INPUT input)
 		float previousVl = PreviousFrameTex.Sample(PreviousFrameSampler, previousScreenPosition).x;
-		float previousDepth = PreviousDepthTex.Sample(PreviousDepthSampler, previousScreenPosition).x;
+		float previousDepth = PreviousDepthTex.Sample(PreviousDepthSampler,
+#	ifndef VR
```
This also may be a flat bug, but I didn't check since dynres isn't toggleable in the gui.

alandtse commented:
Getting reports of ctds for large load orders. Gonna move this to drafts.

Moyse06 commented:
Night sky is very shaky.  Has been this way since August I believe.

alandtse commented:
Not related to VL.

alandtse committed:
fix: require vr address library 0.154.0

alandtse committed:
fix: fix crash of shizof mods at dataloaded

alandtse commented:
Ok, fixed ctds for shizof

alandtse committed:
revert: "feat: add dynres status to state"

This reverts commit c3658ad122e62a787ef1bdeda8a3d6b33414f025.

alandtse committed:
fix: use Util::IsDynamicResolution

alandtse commented:
I'm actually thinking there's no reason to do the CopyResources call. Right now by trying to restore the call when not in dyn res, I'm seeing some tearing as the stencil size actually changes in phases resulting in some artifacts if you're toggling dynres. Looking at renderdoc, I don't see any reason to keep the CopyResources call (which actually was the effect of earlier builds due to a bad boolean check).

alandtse committed:
fix: fix improper depth size when downsampled

alandtse committed:
fix: fix bad rendering with bDepthBufferCulling

With bDepthBufferCulling enabled, culling is incorrect if diskcache is
on until ShaderCache is cleared. This is a stopgap until root cause is
identified.

alandtse commented:
Fixed bDepthBufferCulling bugs:

- Downscaled depth buffer stencil resulting in bad stencil
- Rendering objects improperly culled with disk cache enabled

alandtse committed:
fix(VR): fix VR desync

closes #601

alandtse committed:
revert: "fix(VR): fix VR desync"

This reverts commit 958bc9a94ddfda0a24e1adbc3e2b554e6b1d137a.

alandtse committed:
style: 🎨 apply clang-format changes

alandtse committed:
style: fix doxygen spacing

FlayaN commented on package/Shaders/ISFullScreenVR.hlsl:
```diff
@@ -0,0 +1,69 @@
+	float2 uv = input.TexCoord;  // Get the UV coordinates from input.
+
+	// Convert UV to normalized screen space [-1, 1].
+	float2 normalizedScreenCoord = ConvertUVToNormalizedScreenSpace(uv);
```
`VR::ConvertUVToNormalizedScreenSpace`

alandtse replied:
Le sigh.

alandtse committed:
refactor: use new namespaces

alandtse committed:
fix: fix VR namespace define conflict

FlayaN commented on package/Shaders/ISApplyVolumetricLighting.hlsl:
```diff
@@ -34,8 +35,16 @@ PS_OUTPUT main(PS_INPUT input)
 {
 	PS_OUTPUT psout;
 
+	uint eyeIndex = Stereo::GetEyeIndexFromTexCoord(input.TexCoord);
```
eyeIndex not used? Can also remove `#include "Common/VR.hlsli"` in that case

alandtse committed:
chore: remove unused vr code

alandtse commented:
Can't reproduce any bug with `bUseHiddenAreaMesh`. Looks like a hmd specific bug with upscaler.

alandtse committed:
refactor: fix flat compile errors
